daemon/migration.go

```diff
@@ -0,0 +1,135 @@
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/log"
+	"github.com/containerd/platforms"
+	"github.com/docker/docker/api/types/backend"
+	"github.com/docker/docker/container"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/internal/multierror"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type daemonPlatformReader struct {
+	imageService ImageService
+	content      content.Provider
+}
+
+func (r daemonPlatformReader) ReadPlatformFromConfigByImageManifest(
+	ctx context.Context,
+	desc ocispec.Descriptor,
+) (ocispec.Platform, error) {
+	b, err := content.ReadBlob(ctx, r.content, desc)
+	if err != nil {
+		return ocispec.Platform{}, err
+	}
+
+	var mfst ocispec.Manifest
+	if err := json.Unmarshal(b, &mfst); err != nil {
+		return ocispec.Platform{}, err
+	}
+
+	b, err = content.ReadBlob(ctx, r.content, mfst.Config)
+	if err != nil {
+		return ocispec.Platform{}, err
+	}
+
+	var cfg ocispec.Image
+	if err := json.Unmarshal(b, &cfg); err != nil {
+		return ocispec.Platform{}, err
+	}
+
+	if cfg.Platform.OS == "" || cfg.Platform.Architecture == "" {
+		return ocispec.Platform{}, errors.New("malformed image config")
+	}
+
+	return cfg.Platform, nil
+}
+
+func (r daemonPlatformReader) ReadPlatformFromImage(ctx context.Context, id image.ID) (ocispec.Platform, error) {
+	img, err := r.imageService.GetImage(ctx, id.String(), backend.GetImageOpts{})
+	if err != nil {
+		return ocispec.Platform{}, err
+	}
+
+	return img.Platform(), nil
+}
+
+type platformReader interface {
+	ReadPlatformFromConfigByImageManifest(ctx context.Context, desc ocispec.Descriptor) (ocispec.Platform, error)
+	ReadPlatformFromImage(ctx context.Context, id image.ID) (ocispec.Platform, error)
+}
+
+func migrateContainerOS(ctx context.Context,
+	migration platformReader,
+	ctr *container.Container,
+) {
+	deduced, err := deduceContainerPlatform(ctx, migration, ctr)
+	if err != nil {
+		log.G(ctx).WithFields(log.Fields{
+			"container": ctr.ID,
+			"error":     err,
+		}).Warn("failed to deduce the container architecture, fallback to the host architecture")
+
+		ctr.ImagePlatform = platforms.DefaultSpec()
+		ctr.ImagePlatform.Variant = ""
+		return
+	}
+
+	ctr.ImagePlatform = deduced
+}
+
+func deduceContainerPlatform(
```

Comment:
```suggestion
// deduceContainerPlatform tries to deduce `ctr`'s platform.
// If both `ctr.OS` and `ctr.ImageManifest` are empty, the host's platform
// is assumed.
// Otherwise:
// - `ctr.ImageManifest.Platform` is used, if it exists and is not empty.
// - The platform from the manifest's config is used, if `ctr.ImageManifest` exists
// and we're able to load it's config from the content store.
// - The platform found by loading the image from the image service by ID (using
// `ctr.ImageID`) is used – this looks for the best *present* matching manifest in
// the store.
func deduceContainerPlatform(
```

Reply:
or instead, if we decide not to fallback:
```suggestion
// deduceContainerPlatform tries to deduce `ctr`'s platform.
// If both `ctr.OS` and `ctr.ImageManifest` are empty, it's not possible to deduce
// the container platform and an error is returned.
// Otherwise:
// - `ctr.ImageManifest.Platform` is used, if it exists and is not empty.
// - The platform from the manifest's config is used, if `ctr.ImageManifest` exists
// and we're able to load it's config from the content store.
// - The platform found by loading the image from the image service by ID (using
// `ctr.ImageID`) is used – this looks for the best *present* matching manifest in
// the store.
func deduceContainerPlatform(
```

Reply:
Thanks! Added the comment.